src/rgw/rgw_sal_motr.cc

```diff
@@ -1223,40 +1224,62 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
       keys[0].append(" ");
   }
 
-  rc = store->next_query_by_name(bucket_index_iname, keys, vals, params.prefix,
-                                                                 params.delim);
-  if (rc < 0) {
-    ldpp_dout(dpp, 0) << __func__ << ": ERROR: next_query_by_name failed, rc = " << rc << dendl;
-    return rc;
-  }
-
-  // Check if an extra key was fetched or not
-  if (rc == max) {
-    // One extra key is successfully fetched.
-    results.is_truncated = true;
-    results.next_marker = keys[max - 2];
-    rc--;
-  } else {
-    results.is_truncated = false;
-  }
-
-  // Process the returned pairs to add into ListResults.
-  int i = 0;
-  for (; i < rc; ++i) {
-    if (vals[i].length() == 0) {
-      results.common_prefixes[keys[i]] = true;
-    } else {
-      rgw_bucket_dir_entry ent;
-      auto iter = vals[i].cbegin();
-      ent.decode(iter);
-      std::string null_ref_key = ent.key.name + "/null";
-      if(keys[i] == null_ref_key)
+  results.is_truncated = false;
+  int keycount=0;
+  std::string next_key;
+  while(keycount <= max) {
+    if(!next_key.empty())
+      keys[0].assign(next_key.begin(), next_key.end());
+    rc = store->next_query_by_name(bucket_index_iname, keys, vals, params.prefix,
+                                                                  params.delim);
```

Comment:
Just the formatting suggestions.
```suggestion
  while (keycount <= max) {
    if (!next_key.empty())
      keys[0] = next_key;
    rc = store->next_query_by_name(bucket_index_iname, keys, vals, params.prefix,
                                   params.delim);
```